include/libhal/spi.hpp

```diff
@@ -20,9 +20,275 @@
+       * - Data is shifted out on: rising SCLK
+       * - Data is sampled on: falling SCLK
+       * - CPOL (clock polarity): 0
+       * - CPHA (clock phase): 1
+       */
+      m1,
+
+      /**
+       * @brief spi mode 2
+       *
+       * - Data is shifted out on: rising SCLK, and when CS activates
+       * - Data is sampled on: falling SCLK
+       * - CPOL (clock polarity): 1
+       * - CPHA (clock phase): 0
+       */
+      m2,
+
+      /**
+       * @brief spi mode 3
+       *
+       * - Data is shifted out on: falling SCLK
+       * - Data is sampled on: rising SCLK
+       * - CPOL (clock polarity): 1
+       * - CPHA (clock phase): 1
+       */
+      m3,
+    };
+
+    /**
+     * @brief Best-effort clock rate to set the spi bus to
+     *
+     * This field denotes the clock rate that an application or driver would
+     * like to set the bus to. This clock rate is "best-effort" meaning that the
+     * spi channel will try its best to reach this clock rate but will not
+     * guarantee the exact clock rate.
+     *
+     * The clock rate returned from the API `clock_rate()` will always return a
+     * value less than or equal to this value. If the clock rate is above what
+     * the spi bus can achieve, then the clock rate will be set to the maximum
+     * the spi bus can manage.
+     */
+    u32 clock_rate = 100_kHz;
+
+    /**
+     * @brief Mode select field
+     *
+     * Use this to select the spi bus mode.
+     */
+    mode select = mode::m0;
+  };
+
+  /**
+   * @brief Set the spi settings for this channel
+   *
+   * The settings passed to this API will be stored by the implementation of
+   * this interface for use when `chip_select(true)` is called.
+   *
+   * @param p_settings - settings to configure the spi channel
+   */
+  void configure(settings const& p_settings);
+
+  /**
+   * @brief Return the clock rate that the SPI channel will operate at when a
+   * transfer occurs.
+   *
+   * The the clock rate returned is the clock rate the spi channel will be
+   * configured to when `chip_select(true)` and is not the current clock rate of
+   * the bus.
+   *
+   * The clock rate reported here is guaranteed to be less than or equal to the
+   * value `clock_rate` provided to the configure API. This function can be
+   * called by the program to determine the delta between the requested
+   * frequency passed to `configure()` and the actual clock frequency the bus
+   * will be set to.
+   *
+   * @return u32 - the approximate clock rate of this spi channel when
+   * `chip_select(true)` is called.
+   */
+  u32 clock_rate();
+
+  /**
+   * @brief Send and receive data between a selected device on the spi bus.
+   *
+   * This function will block until the entire transfer is finished. This API
+   * will only work if `chip_select(true)` was called prior, otherwise, this API
+   * does nothing.
+   *
+   * TODO(kammce): consider throwing an exception if `chip_select(false)` when
+   * performing a transfer
```

Comment:
do not submit, remove comment on

Reply:
I finally spotted the answer to this in the docs in libhal.github.io. The answer is that calling transfer also asserts chip select but only during the duration of the transaction. transaction will block on the bus lock until the bus is available.